crates/consensus-logic/src/sync_manager.rs

```diff
@@ -134,3 +172,74 @@ pub fn start_sync_tasks<E: ExecEngineCtl + Sync + Send + 'static>(
         status_channel,
     })
 }
+
+fn spawn_exec_worker<E: ExecEngineCtl + Sync + Send + 'static>(
+    shutdown: ShutdownGuard,
+    handle: Handle,
+    storage: Arc<NodeStorage>,
+    status_channel: StatusChannel,
+    engine: Arc<E>,
+    exec_rx: ExecCtlInput,
+) -> anyhow::Result<()> {
+    info!("waiting until genesis");
+    let init_state = handle.block_on(status_channel.wait_until_genesis())?;
+    let cur_tip = match init_state.get_declared_final_epoch().cloned() {
+        Some(epoch) => epoch.to_block_commitment(),
+        None => L2BlockCommitment::new(
+            0,
+            *init_state.sync().expect("after genesis").genesis_blkid(),
+        ),
+    };
+
+    let blkid = *cur_tip.blkid();
+    info!(%blkid, "starting exec worker");
+
+    let exec_env_id = ();
+    let state = ExecWorkerState::new(engine, exec_env_id, cur_tip, cur_tip);
+    let ctx = ExecWorkerCtx::new(storage.l2().clone());
+    exec_worker_task(shutdown, state, exec_rx, &ctx)?;
+    Ok(())
+}
+
+fn spawn_chain_worker(
+    shutdown: ShutdownGuard,
+    handle: Handle,
+    storage: Arc<NodeStorage>,
+    status_channel: StatusChannel,
+    params: Arc<Params>,
+    exect_ctl_handle: ExecCtlHandle,
+    chain_msg_rx: mpsc::Receiver<ChainWorkerMessage>,
+) -> anyhow::Result<()> {
+    info!("waiting until genesis");
+    let init_state = handle.block_on(status_channel.wait_until_genesis())?;
+    let cur_tip = match init_state.get_declared_final_epoch().cloned() {
+        Some(epoch) => epoch.to_block_commitment(),
+        None => L2BlockCommitment::new(
+            0,
+            *init_state.sync().expect("after genesis").genesis_blkid(),
+        ),
+    };
+
+    let blkid = *cur_tip.blkid();
+    info!(%blkid, "starting chain worker");
+
+    let context = ChainWorkerCtx::new(
+        storage.l2().clone(),
+        storage.chainstate().clone(),
+        storage.checkpoint().clone(),
+        0, // FIXME: Not sure what this is
+    );
+    let chain_exec = ChainExecutor::new(params.rollup().clone());
+    let shared = Arc::new(Mutex::new(WorkerShared::default()));
+    let state = strata_chain_worker::init_worker_state(
+        shared.clone(),
+        context,
+        chain_exec,
+        exect_ctl_handle,
+        cur_tip,
+    )?;
+    let input = ChainWorkerInput::new(shared.clone(), chain_msg_rx);
```

Comment:
There's a lot of init logic happening here that belongs in the chain worker modules.  Like the `WorkerShared` struct is solidly an implementation detail that should not even be exported.  The ideal way to construct these would have been to have some function that spawns the worker in a new thread and returns the `ChainWorkerHandle`.  Then this handle would be passed to the FCM when setting that up.

I realize that maybe there's some constraints on needing to wait for genesis to happen that makes that more complicated.  It would be better to encapsulate this complexity and shifting it down into the worker task somehow, instead of exposing it.